include/swift/SIL/SILDifferentiabilityWitness.h

```diff
@@ -0,0 +1,156 @@
+      public SILAllocated<SILDifferentiabilityWitness>
+{
+private:
+  /// The module which contains the differentiability witness.
+  SILModule &module;
+  /// The linkage of the differentiability witness.
+  SILLinkage linkage;
+  /// The original function.
+  SILFunction *originalFunction;
+  /// The parameter indices.
+  IndexSubset *parameterIndices;
+  /// The result indices.
+  IndexSubset *resultIndices;
+  /// The derivative generic signature (optional).
+  GenericSignature *derivativeGenericSignature;
+  /// The JVP (Jacobian-vector products) derivative function.
+  SILFunction *jvp;
+  /// The VJP (vector-Jacobian products) derivative function.
+  SILFunction *vjp;
+  /// Whether or not this differentiability witness is serialized, which allows
+  /// devirtualization from another module.
+  bool serialized;
+  /// The AST `@differentiable` or `@differentiating` attribute from which the
+  /// differentiability witness is generated. Used for diagnostics.
+  /// Null if the differentiability witness is parsed from SIL or if it is
+  /// deserialized.
+  DeclAttribute *attribute = nullptr;
+
+  static AutoDiffConfig *
+  getAutoDiffConfig(SILModule &module, IndexSubset *parameterIndices,
+                    IndexSubset *resultIndices,
+                    GenericSignature *derivativeGenSig);
+
+  SILDifferentiabilityWitness(SILModule &module, SILLinkage linkage,
+                              SILFunction *originalFunction,
+                              IndexSubset *parameterIndices,
+                              IndexSubset *resultIndices,
+                              GenericSignature *derivativeGenSig,
+                              SILFunction *jvp, SILFunction *vjp,
+                              bool isSerialized, DeclAttribute *attribute)
+    : module(module), linkage(linkage), originalFunction(originalFunction),
+      parameterIndices(parameterIndices), resultIndices(resultIndices),
+      derivativeGenericSignature(derivativeGenSig), jvp(jvp), vjp(vjp),
+      serialized(isSerialized), attribute(attribute) {}
+
+public:
+  static SILDifferentiabilityWitness *create(
+      SILModule &module, SILLinkage linkage, SILFunction *originalFunction,
+      IndexSubset *parameterIndices, IndexSubset *resultIndices,
+      GenericSignature *derivativeGenSig, SILFunction *jvp, SILFunction *vjp,
+      bool isSerialized, DeclAttribute *attribute = nullptr);
+
+  SILDifferentiabilityWitnessKey getKey() const;
+  SILModule &getModule() const { return module; }
+  SILLinkage getLinkage() const { return linkage; }
+  SILFunction *getOriginalFunction() const { return originalFunction; }
+  IndexSubset *getParameterIndices() const {
+    return parameterIndices;
+  }
+  IndexSubset *getResultIndices() const {
+    return resultIndices;
+  }
+  GenericSignature *getDerivativeGenericSignature() const {
+    return derivativeGenericSignature;
+  }
+  SILFunction *getJVP() const { return jvp; }
```

Comment:
It seems a bit odd to not have `setJVP` and `setVJP` when there's `setDerivative`.

Reply:
Sorry, `setDerivative` isn't used in this PR but is used in my next patch. It's more convenient for callers than `setJVP` and `setVJP`, since callers have to do a `switch` over derivative kind.

I'll remove the utility from this patch since it's not yet used.

Reply:
Actually, I decided to add `setJVP` and `setVJP`. Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.